packages/proxy/lib/http/util/prerequests.ts

```diff
@@ -12,101 +11,86 @@ const metrics: any = {
   browserPreRequestsReceived: 0,
   proxyRequestsReceived: 0,
   immediatelyMatchedRequests: 0,
-  eventuallyReceivedPreRequest: [],
-  neverReceivedPreRequest: [],
+  unmatchedRequests: 0,
+  unmatchedPreRequests: 0,
 }
 
 process.once('exit', () => {
   debug('metrics: %o', metrics)
 })
 
-function removeOne<T> (a: Array<T>, predicate: (v: T) => boolean): T | void {
-  for (let i = a.length - 1; i >= 0; i--) {
-    const v = a[i]
-
-    if (predicate(v)) {
-      a.splice(i, 1)
-
-      return v
-    }
-  }
-}
+export type GetPreRequestCb = (browserPreRequest?: BrowserPreRequest) => void
 
-function matches (preRequest: BrowserPreRequest, req: Pick<CypressIncomingRequest, 'proxiedUrl' | 'method'>) {
-  return preRequest.method === req.method && preRequest.url === req.proxiedUrl
+type PendingRequest = {
+  ctxDebug
+  callback: GetPreRequestCb
+  timeout: ReturnType<typeof setTimeout>
 }
 
-export type GetPreRequestCb = (browserPreRequest?: BrowserPreRequest) => void
-
 export class PreRequests {
-  pendingBrowserPreRequests: Array<BrowserPreRequest> = []
-  requestsPendingPreRequestCbs: Array<{
-    cb: (browserPreRequest: BrowserPreRequest) => void
-    method: string
-    proxiedUrl: string
-  }> = []
-
-  get (req: CypressIncomingRequest, ctxDebug, cb: GetPreRequestCb) {
-    metrics.proxyRequestsReceived++
-
-    const pendingBrowserPreRequest = removeOne(this.pendingBrowserPreRequests, (browserPreRequest) => {
-      return matches(browserPreRequest, req)
-    })
-
-    if (pendingBrowserPreRequest) {
-      metrics.immediatelyMatchedRequests++
-
-      ctxDebug('matches pending pre-request %o', pendingBrowserPreRequest)
+  requestTimeout: number
+  pendingPreRequests: Record<string, BrowserPreRequest> = {}
+  pendingRequests: Record<string, PendingRequest> = {}
+  prerequestTimeouts: Record<string, number> = {}
+
+  constructor (requestTimeout = 500) {
+    this.requestTimeout = requestTimeout
+    setInterval(() => {
+      const now = Date.now()
+
+      Object.entries(this.prerequestTimeouts).forEach(([key, timeout]) => {
+        if (timeout < now) {
+          debugVerbose('timed out unmatched pre-request %o', this.pendingPreRequests[key])
```

Comment:
to align with previous verbiage, should we say something like:

```suggestion
          debugVerbose('timed out after %sms, continuing unmatched pre-request %o with an empty pre-request field!', this.requestTimeout, this.pendingPreRequests[key])
```

Reply:
I wasn't fond of the previous style, but your comment made me go through and redo all the debug messages in here. Let me know what you think of the new versions.

Reply:
The new ones are great! Love the clear messaging.